[TF] TF-496: Conditionally conform SIMD types to Differentiable

bartchr808 opened the pull request:
Implements this [JIRA ticket](https://bugs.swift.org/browse/TF-496) which makes SIMD vectors differentiable.

Also here is a link to the [Swift Forums](https://forums.swift.org/t/simd-conformance-to-additivearithmetic/24492/6) where some of this was discussed. See the forum post for any context regarding the changes.

There is also the change to the `-=`/`+=` operator. As background, we need `SIMD` to conform to the `AdditiveArithmetic ` protocol which already provides a conditional definition of the `-=` and `+=` operator. So to avoid the "dual default conditional definition protocol" issue, I removed the default implementation in SIMD for now on our branch, but we will fix it when we pitch it to get it merged into master.

Also, I had to remove `@_alwaysEmitIntoClient` from `sum()` due to a bug (filed as [TF-545](https://bugs.swift.org/browse/TF-545)).

And finally, there are custom VJPs that I defined that don't actually need to be defined, as the body of the original methods define methods that I already marked as differentiable. But the bug [TF-547](https://bugs.swift.org/browse/TF-547) is blocking me on that.

bartchr808 committed:
Initial broken version.

bartchr808 committed:
Add VectorView.

bartchr808 committed:
WIP while waiting for swift forums answer.

bartchr808 committed:
WIP: removed from AdditiveArithmetic, kept in CIMD.

bartchr808 committed:
Remove += from SIMD protocol, move it to Struct SIMDn.

bartchr808 committed:
Have SIMDn conform to Differentiable.

bartchr808 commented:
Forgot to remove the `zero` property....will push change soon.

bartchr808 committed:
Forgot to remove zero, already implemented in SIMDVector.swift

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +185,19 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
   }
 }
 
+extension SIMD${n} : AdditiveArithmetic where Scalar: FloatingPoint {}
```
```suggestion
extension SIMD${n} : AdditiveArithmetic where Scalar : FloatingPoint {}
```

bartchr808 committed:
PR fix: spacing.

bartchr808 commented:
@stephentyrone I'm not sure what you and the the Swift Core team thinks as to whether this is a bug fix or a new feature that would require a proposal for, but I just realized that we could create a separate PR just for the `-=`/`+=` part and merge that into `master`.

Additionally, for the `+=`/`-=` fix of my PR, which do you think would be better (or possibly even an alternate idea if you have one):
- My current PR where I removed the operators from the `SIMD` protocol, keeping the one in `AdditiveArithmetic`
- The complete reverse where we keep the `SIMD` protocol conditional definition, and remove the conditional definition in `AdditiveArithmetic`

bartchr808 committed:
Start making + differentiable

bartchr808 committed:
Get + and - differentiable.

bartchr808 committed:
WIP

bartchr808 committed:
Make all necessary operators differentiable.

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +187,34 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
+  where Scalar : Differentiable & BinaryFloatingPoint,
+        Scalar.CotangentVector : BinaryFloatingPoint {
+  public func _vjpSubscript(index: Int) ->
+  (Scalar, (Scalar.CotangentVector) -> CotangentVector)
```
Indent by 2 and put the `->` on this line.
```suggestion
  -> (Scalar, (Scalar.CotangentVector) -> CotangentVector)
```

bartchr808 replied:
To verify:
```swift
public func _vjpSubscript(index: Int)
  -> (Scalar, (Scalar.CotangentVector) -> CotangentVector)
```

rxwei replied:
Yes. VJP functions should not be public though. When the original function is public, a VJP function should be internal. When it's internal, if the original function has `@inlinable`, the VJP function should be `@inlinable`, otherwise the VJP function should be `@usableFromInline`.

bartchr808 replied:
Ahh okay. But for let's say `_vjpSum`, the original function (in the protocol) is marked as public and `@inlinable`. How should `_vjpSum` be marked? Looking at `Tensor`, its `_vjpSum` is just `@inlinable` in a regular extension

rxwei replied:
If the original function is `@inlinable`, the VJP should also be `@inlinable`. In all cases where the original function is public, the VJP should be internal.

rxwei replied:
But for `SIMD.sum()`, it's not `@inlinable`, so the VJP should be `@usableFromInline`.

bartchr808 replied:
Ahh okay thanks!

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +187,34 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
+        Scalar.CotangentVector : BinaryFloatingPoint {
+  public func _vjpSubscript(index: Int) ->
+  (Scalar, (Scalar.CotangentVector) -> CotangentVector)
+  {
```
Put `{` at the end of the previous line.

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +187,34 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
+      zeros[index] = Scalar(v)
+      return zeros
+    }
+    )
```
Put `)` right after `}` on the previous line.

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +187,34 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
+  public func _vjpSubscript(index: Int) ->
+  (Scalar, (Scalar.CotangentVector) -> CotangentVector)
+  {
+    return (self[index], { (v: Scalar.CotangentVector) in
```
I don't think the type signature is needed here. Did the compiler suggest otherwise?

```suggestion
    return (self[index], { v in
```

bartchr808 replied:
At some point it was asking me for it either here in the struct or in the `SIMD` protocol where it gave me an error along the lines of `cannot convert return expression of type '(Self, (_) -> _)' to return type '(Self, (Self.CotangentVector) -> (Self.Scalar, Self.CotangentVector))'`, but I think this ended up being fixed when I corrected the protocol refinements. So I actually can remove these all now!

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpDivide(
+    lhs: Scalar, rhs: Self
+    ) -> (Self, (CotangentVector) -> (Scalar.CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpMultiply(
+    lhs: Scalar, rhs: Self
+    ) -> (Self, (CotangentVector) -> (Scalar.CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpDivide(
+    lhs: Self, rhs: Scalar
+    ) -> (Self, (CotangentVector) -> (CotangentVector, Scalar.CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpMultiply(
+    lhs: Self, rhs: Scalar
+    ) -> (Self, (CotangentVector) -> (CotangentVector, Scalar.CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  static func _vjpMultiply(
+    lhs: Self, rhs: Scalar
+    ) -> (Self, (CotangentVector) -> (CotangentVector, Scalar.CotangentVector)) {
+    return (lhs * rhs, { (v: CotangentVector) in
```
Hmm, I don't think a type signature is required for any pullback argument. If it's part of the `return` statement, it should be inferred.
```suggestion
    return (lhs * rhs, { v in
```

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpSubtract(
+    lhs: Self, rhs: Scalar
+    ) -> (Self, (CotangentVector) -> (CotangentVector, Scalar.CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpAdd(
+    lhs: Self, rhs: Scalar
+    ) -> (Self, (CotangentVector) -> (CotangentVector, Scalar.CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpSubtract(
+    lhs: Scalar, rhs: Self
+    ) -> (Self, (CotangentVector) -> (Scalar.CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpAdd(
+    lhs: Scalar, rhs: Self
+    ) -> (Self, (CotangentVector) -> (Scalar.CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpDivide(
+    lhs: Self, rhs: Self
+    ) -> (Self, (CotangentVector) -> (CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpMultiply(
+    lhs: Self, rhs: Self
+    ) -> (Self, (CotangentVector) -> (CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+extension SIMD
+  where Self: Differentiable,
+        CotangentVector: SIMD,
+        // error: generic parameter 'Self' could not be inferred: return (lhs * rhs,...
```
For merging into stdlib, these comments should be removed.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpSubtract(
+    lhs: Self, rhs: Self
+    ) -> (Self, (CotangentVector) -> (CotangentVector, CotangentVector)) {
```
Un-indent.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+  @inlinable
+  static func _vjpAdd(
+    lhs: Self, rhs: Self
+    ) -> (Self, (CotangentVector) -> (CotangentVector, CotangentVector)) {
```
Un-indent.

bartchr808 replied:
Un-indent once so it's inline with `static` right?

rxwei replied:
Yes

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -781,20 +781,35 @@ extension SIMD where Scalar: FixedWidthInteger {
 //  be replaced with @_semantics to lower directly to vector IR nodes.
 extension SIMD where Scalar: FloatingPoint {
   @_transparent
+  @differentiable(vjp: _vjpAdd(lhs:rhs:)
```
Could you add a `// SWIFT_ENABLE_TENSORFLOW` in the beginning of every diff blob? We use this to determine whether something is a S4TF-specific change.

bartchr808 replied:
Done!

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1461,154 @@ where T: SIMD, T.Scalar: FloatingPoint {
+        CotangentVector == Self {
+  @usableFromInline
+  func _vjpSum() -> (Scalar, (Scalar.CotangentVector) -> CotangentVector) {
+    return (sum(), { (v: Scalar.CotangentVector)  in Self.init(repeating: Scalar(v)) })
```
Omit `.init` and prefer the initialization syntax.

```suggestion
    return (sum(), { v in Self(repeating: Scalar(v)) })
```

bartchr808 committed:
Address 1st batch of PR comments.

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -44,7 +44,9 @@ public struct SIMD${n}<Scalar>: SIMD where Scalar: SIMDScalar {
     _storage = Scalar.SIMD${storageN}Storage()
   }
 
-  /// Accesses the scalar at the specified position.
```
Don't delete existing doc comments :)

bartchr808 replied:
Oh oops!

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +187,32 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
   }
 }
 
+extension SIMD${n} : AdditiveArithmetic where Scalar : FloatingPoint {}
```
Add `// SWIFT_ENABLE_TENSORFLOW` above.

rxwei commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -44,7 +44,9 @@ public struct SIMD${n}<Scalar>: SIMD where Scalar: SIMDScalar {
     _storage = Scalar.SIMD${storageN}Storage()
   }
 
-  /// Accesses the scalar at the specified position.
```
Add `// SWIFT_ENABLE_TENSORFLOW` above.

rxwei commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1473,152 @@ where T: SIMD, T.Scalar: FloatingPoint {
   }
   return result
 }
+
```
Add `// SWIFT_ENABLE_TENSORFLOW` above.

bartchr808 replied:
I should do this for each extension I added right?

bartchr808 committed:
Fix due to changes in PR #24825.

bartchr808 committed:
Address 2nd batch of PR comments.

bartchr808 committed:
Make _vjpSubscript internal and inline.

bartchr808 commented:
Seems like this may have a bug, filed it [here](https://bugs.swift.org/browse/TF-508)

bartchr808 committed:
WIP: test for + doesn't seem to work, filed JIRA bug.

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

bartchr808 committed:
WIP: add back functions

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

bartchr808 committed:
Make negation diff & start fixing tests.

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

bartchr808 committed:
Fix bugs in code and add tests

TODO:
- Determine whether sum() needs to be differentiated
- See what other inits I can differentiate
- Remove some of the custom VJPs now that init(repeating:) is differentiable

bartchr808 commented:
In case you see that I pushed commits to this PR, Ignore re-reviewing this PR for now while I clean up some stuff. Pushed it so it's accessible from desktop and laptop for me. Still need to do three things:

- Determine whether sum() needs to be differentiated
- See what other inits I can differentiate
- Remove some of the custom VJPs now that init(repeating:) is differentiable

bartchr808 commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -843,6 +874,13 @@ extension SIMD where Scalar: FloatingPoint {
   
   /// Returns the sum of the scalars in the vector.
   @_alwaysEmitIntoClient
```
Need to look into this, but I think this attribute is what is preventing me from differentiating the sum()

bartchr808 committed:
Get sum() differentiable.

bartchr808 committed:
Cleanup and additional init(repeating:) test.

bartchr808 committed:
Fix up AdditiveArithmetic conformance.

bartchr808 commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1157,60 +1197,112 @@ extension SIMD where Scalar: FixedWidthInteger {
   }
   
-  @_transparent
-  public static func +=(lhs: inout Self, rhs: Self) {
```
In order to make SIMD conform to AdditiveArithmetic, I had to remove both of these implementations as two protocols cannot both have default implementations. Let me know if this is okay, or if there is something else we should do.

bartchr808 commented:
@swift-ci please test tensorflow

bartchr808 commented:
@swift-ci please test tensorflow

bartchr808 commented:
TODO: getting assertion failure if I want to eliminate redundant custom VJPs: https://bugs.swift.org/browse/TF-547

bartchr808 committed:
Merge branch 'tensorflow' into TF-496-conform-SIMD-types-differentiable

bartchr808 committed:
Code cleanup and additonal tests.

bartchr808 committed:
White space.

dan-zheng commented on test/AutoDiff/SIMD.swift:
```diff
@@ -0,0 +1,194 @@
+import Glibc
+#endif
+
+var SIMDTests = TestSuite("SIMD")
```
A testing idea that needn't block progress:

Currently, all tests seem to operate on `SIMD<Float>`. It would be nice to test differentiation of generic functions over `SIMD`, with the appropriate generic constraints. Example:
```swift
func testInit<Scalar, SIMDType>(x: Scalar) -> SIMDType
  where SIMDType : Differentiable,
        SIMDType.TangentVector : SIMD,
        SIMDType.Scalar == Scalar,
        Scalar : BinaryFloatingPoint & Differentiable,
        Scalar.TangentVector : BinaryFloatingPoint,
        SIMDType.TangentVector == SIMDType {
  return SIMDType.init(repeating: x)
}
```

Defining generic functions also allows us to test multiple `SIMD` and `Scalar` type combinations, e.g. `SIMD2<Double>`.

bartchr808 replied:
Good idea! I can definitely a couple tests regarding generics. It would be especially good to do it given the way `SIMD` is implemented with a protocol and then structs.

bartchr808 replied:
@dan-zheng the constraints on the vjpInit are the following:
```swift
extension SIMD
  where Self : Differentiable,
        Self.TangentVector : SIMD,
        Scalar : BinaryFloatingPoint & Differentiable,
        Self.TangentVector == Self,
        Scalar.TangentVector == Scalar {
  @usableFromInline
  static func _vjpInit(repeating value: Scalar)
    -> (Self, (TangentVector) -> Scalar.TangentVector) {
      return (Self(repeating: value), { v in v.sum() })
  }
}
```
Thus I think a generic init function would look like this:
```swift
func testInit<Scalar, SIMDType>(x: Scalar) -> SIMDType
  where SIMDType : Differentiable,
        SIMDType.TangentVector : SIMD,
        Scalar : BinaryFloatingPoint & Differentiable,
        SIMDType.TangentVector == SIMDType,
        Scalar.TangentVector == Scalar {
  return SIMDType.init(repeating: x)
}
```
However I'm getting an error saying it's expecting me to use the `init(from decoder: Decoder)` constructor:
```
simd.swift:7:27: error: incorrect argument label in call (have 'repeating:', expected 'from:')
      return SIMDType.init(repeating: x)
                          ^~~~~~~~~~
                           from
```
Is there something wrong in my constraints, or is this a bug?

bartchr808 replied:
@dan-zheng quick ping when you have a minute

dan-zheng replied:
The `SIMDType.Scalar == Scalar` constraint is missing.
(The original program I shared does have that constraint.)

bartchr808 replied:
Either there is a bug or there is another constraint missing (looking right now), but I can't mark the function you posted as differentiable:
```swift
@differentiable
func testInit<Scalar, SIMDType>(x: Scalar) -> SIMDType
  where SIMDType : Differentiable,
        SIMDType.TangentVector : SIMD,
        SIMDType.Scalar == Scalar,
        Scalar : BinaryFloatingPoint & Differentiable,
        Scalar.TangentVector : BinaryFloatingPoint,
        SIMDType.TangentVector == SIMDType {
  return SIMDType.init(repeating: x)
}
```

```
simd.swift:24:19: error: expression is not differentiable
  return SIMDType.init(repeating: x)
                  ^
simd.swift:24:19: note: function call is not differentiable because generic requirements are not met
  return SIMDType.init(repeating: x)
                  ^
```

bartchr808 replied:
That's why my function had different constraints, as I was running into this issue and was messing around with the constraints.

bartchr808 replied:
Added generic tests in [`d73f3da`](https://github.com/apple/swift/pull/24786/commits/d73f3dab585cb17ea128e215a634bf2c8d6cf985).

dan-zheng reviewed (approved):
LGTM!

bartchr808 commented on stdlib/public/core/SIMDVectorTypes.swift.gyb:
```diff
@@ -185,6 +189,33 @@ extension SIMD${n} where Scalar: BinaryFloatingPoint {
+  -> (Scalar, (Scalar.TangentVector) -> TangentVector) {
+    return (self[index], { v in
+      var zeros = Self.zero
+      zeros[index] = Scalar(v)
```
v is of type `Scalar.TangentVector` while `zeros` is of type `SIMD3<Scalar>`, and since the only constraint on `Scalar.TangentVector` is `Scalar.TangentVector : BinaryFloatingPoint` which `Scalar` also conforms to, I have to cast it or else I I get the error `cannot assign value of type 'Scalar.TangentVector' to type 'Scalar'`. I could add the more constraining requirement of `Scalar.TangentVector == Scalar`, but I think doing this cast is better. Let me know what you think!

dan-zheng replied:
cc @rxwei regarding whether to add `Scalar.TangentVector == Scalar` constraint.
I agree that not adding the constraint is better.

bartchr808 commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -1407,3 +1499,159 @@ where T: SIMD, T.Scalar: FloatingPoint {
+        TangentVector == Self {
+  @inlinable
+  func _vjpSum() -> (Scalar, (Scalar.TangentVector) -> TangentVector) {
+    return (sum(), { v in Self(repeating: Scalar(v)) })
```
Similar to my comment below about the `Scalar` initializer call, I would need the more strict requirement of `Scalar.TangentVector == Scalar` instead of `Scalar.TangentVector : BinaryFloatingPoint`, or else I get the error
```
error: cannot convert return expression of type '(Self.Scalar) -> Self' to return type '(_) -> _'
    return (sum(), { v in Self(repeating: v) })
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

bartchr808 committed:
Add generic tests.

bartchr808 commented:
@swift-ci please test tensorflow

dan-zheng reviewed (approved):
Generic SIMD differentiation tests look great!

bartchr808 commented:
@dan-zheng macOS tests are still going 16 hours later, is this a known problem currently? Or is this the linux passing good enough to merge in?

rxwei reviewed (approved):
Great work Bart!

bartchr808 commented:
Thanks @dan-zheng and @rxwei ! 😄
